{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Translate GOLD study, project, and biosample data into json.\n",
    "The notebooks demostrates how to translate study, project, and biosample data from the GOLD database into json that conforms with the [NMDC schema](https://github.com/microbiomedata/nmdc-metadata/blob/schema-draft/README.md).  \n",
    "Before doing the translation it is important that you have an up to date `nmdc.py` file in the `lib` directory.  \n",
    "\n",
    "The python modules for running the notebook are in the `requirements.txt` file.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pds\n",
    "import jsonasobj\n",
    "import lib.nmdc as nmdc\n",
    "import lib.data_operations as dop\n",
    "from pandasql import sqldf\n",
    "\n",
    "def pysqldf(q):\n",
    "    return sqldf(q, globals())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load tables (i.e., tab delimited files) from nmdc zip file\n",
    "The NMDC data is currently stored in a zip file. Instead of unzipping the file, simply use the `zipfile` library to load the `study`, `project`, `contact`, `project_biosample`, and `biosample` tables (stored as tab-delimited files). The `project_biosample` table is needed as a cross-linking table between `project` and `biosample`. The `contact` table contains information about principal investigators.\n",
    "\n",
    "The code for unzipping and creating the dataframe is found in the `make_dataframe` function. As part of the dataframe creation process, the column names are lower cased and spaces are replaced with underscored. I find it helpful to have some standarization on column names when doing data wrangling. This behavior can be overridden if you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "study = dop.make_dataframe(\"export.sql/STUDY_DATA_TABLE.dsv\", file_archive_name=\"data/nmdc-version2.zip\")\n",
    "contact = dop.make_dataframe(\"export.sql/CONTACT_DATA_TABLE.dsv\", file_archive_name=\"data/nmdc-version2.zip\")\n",
    "project = dop.make_dataframe(\"export.sql/PROJECT_DATA_TABLE.dsv\", file_archive_name=\"data/nmdc-version2.zip\")\n",
    "project_biosample = dop.make_dataframe(\"export.sql/PROJECT_BIOSAMPLE_DATA_TABLE.dsv\", file_archive_name=\"data/nmdc-version2.zip\")\n",
    "biosample = dop.make_dataframe(\"export.sql/BIOSAMPLE_DATA_TABLE.dsv\", file_archive_name=\"data/nmdc-version2.zip\")\n",
    "proposals = dop.make_dataframe(\"data/JGI-EMSL-FICUS-proposals.fnl.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subset tables to records where active = 'Yes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = \"\"\"\n",
    "select \n",
    "    *\n",
    "from\n",
    "    study\n",
    "where\n",
    "    active = 'Yes'\n",
    "\"\"\"\n",
    "study = sqldf(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = \"\"\"\n",
    "select \n",
    "    *\n",
    "from\n",
    "    project\n",
    "where\n",
    "    active = 'Yes'\n",
    "\"\"\"\n",
    "project = sqldf(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = \"\"\"\n",
    "select \n",
    "    *\n",
    "from\n",
    "    biosample\n",
    "where\n",
    "    active = 'Yes'\n",
    "\"\"\"\n",
    "biosample = sqldf(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# biosample.head() # peek at data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build json data files\n",
    "The json data files are build using a general approach:\n",
    "1. Create a pandas dataframe (often using SQL syntax) to be translated.\n",
    "2. Transform the dataframe into a dictionary (these variables end with '_dictdf')\n",
    "3. Define a list of field names whose names and values will be translated into characteristics within an annotation object.\n",
    "4. Pass the dataframe dictionary and characteristices list to the `make_json_string_list` method. This method returns a list of json ojbects each of which has been converted to a string.\n",
    "5. Save the json string to file using `save_json_string_list`.\n",
    "\n",
    "**Note:** Currently, I am using the GOLD IDs as idenifiers. This need to changed to IRIs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build study json\n",
    "* Create a subset of the study table using the FICUS gold_ids specified in [JGI-EMSL-FICUS-proposals.fnl.xlxs](https://docs.google.com/spreadsheets/d/1sowTCYooDrOMq0ErD4s3xtgH3PLoxwa7/edit#gid=1363834365).\n",
    "* Follow approach for building json data files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = \"\"\"\n",
    "select \n",
    "    study.*, contact.name as principal_investigator_name, proposals.doi\n",
    "from\n",
    "    study\n",
    "left join\n",
    "    contact\n",
    "on\n",
    "    study.contact_id = contact.contact_id\n",
    "left join\n",
    "    proposals\n",
    "on\n",
    "    study.gold_id = proposals.gold_study\n",
    "where\n",
    "    study.gold_id in \n",
    "      ('Gs0110115', 'Gs0110132', 'Gs0112340', 'Gs0114675', 'Gs0128849', 'Gs0130354', \n",
    "       'Gs0114298', 'Gs0114663', 'Gs0120351', 'Gs0134277', 'Gs0133461', 'Gs0135152', 'Gs0135149')\n",
    "\"\"\"\n",
    "study = sqldf(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>study_name</th>\n",
       "      <th>its_proposal_id</th>\n",
       "      <th>ncbi_project_id</th>\n",
       "      <th>add_date</th>\n",
       "      <th>mod_date</th>\n",
       "      <th>contact_id</th>\n",
       "      <th>last_mod_by</th>\n",
       "      <th>is_public</th>\n",
       "      <th>gpts_proposal_id</th>\n",
       "      <th>...</th>\n",
       "      <th>public_dap_count</th>\n",
       "      <th>admin_dap_count</th>\n",
       "      <th>public_biosample_count</th>\n",
       "      <th>admin_biosample_count</th>\n",
       "      <th>is_geba</th>\n",
       "      <th>is_hmp</th>\n",
       "      <th>ecosystem_path_id</th>\n",
       "      <th>is_test</th>\n",
       "      <th>principal_investigator_name</th>\n",
       "      <th>doi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>114663</td>\n",
       "      <td>Groundwater microbial communities from the Col...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>12-MAY-15 12.00.00.000000000 AM</td>\n",
       "      <td>08-JAN-16 12.00.00.000000000 AM</td>\n",
       "      <td>113389</td>\n",
       "      <td>102151.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>4165</td>\n",
       "      <td>No</td>\n",
       "      <td>James Stegen</td>\n",
       "      <td>10.25585/1487765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>114298</td>\n",
       "      <td>Permafrost soil microbial communities from the...</td>\n",
       "      <td>1782.0</td>\n",
       "      <td>None</td>\n",
       "      <td>05-FEB-15 12.00.00.000000000 AM</td>\n",
       "      <td>None</td>\n",
       "      <td>102898</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>4234</td>\n",
       "      <td>No</td>\n",
       "      <td>Byron Crump</td>\n",
       "      <td>10.25585/1487766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110115</td>\n",
       "      <td>Avena fatua rhizosphere microbial communities ...</td>\n",
       "      <td>1472.0</td>\n",
       "      <td>None</td>\n",
       "      <td>11-NOV-13 12.00.00.000000000 AM</td>\n",
       "      <td>29-MAY-14 12.00.00.000000000 AM</td>\n",
       "      <td>108484</td>\n",
       "      <td>103865.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>3948</td>\n",
       "      <td>No</td>\n",
       "      <td>Mary Firestone</td>\n",
       "      <td>10.25585/1487760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110132</td>\n",
       "      <td>Cyanobacterial communities from the Joint Geno...</td>\n",
       "      <td>1393.0</td>\n",
       "      <td>None</td>\n",
       "      <td>13-DEC-13 12.00.00.000000000 AM</td>\n",
       "      <td>07-FEB-14 12.00.00.000000000 AM</td>\n",
       "      <td>100043</td>\n",
       "      <td>103865.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>4318</td>\n",
       "      <td>No</td>\n",
       "      <td>Matthias Hess</td>\n",
       "      <td>10.25585/1487758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>112340</td>\n",
       "      <td>Thawing permafrost microbial communities from ...</td>\n",
       "      <td>1778.0</td>\n",
       "      <td>None</td>\n",
       "      <td>24-OCT-14 12.00.00.000000000 AM</td>\n",
       "      <td>31-OCT-14 12.00.00.000000000 AM</td>\n",
       "      <td>110735</td>\n",
       "      <td>105383.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>4206</td>\n",
       "      <td>No</td>\n",
       "      <td>Virginia Rich</td>\n",
       "      <td>10.25585/1487764</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   study_id                                         study_name  \\\n",
       "0    114663  Groundwater microbial communities from the Col...   \n",
       "1    114298  Permafrost soil microbial communities from the...   \n",
       "2    110115  Avena fatua rhizosphere microbial communities ...   \n",
       "3    110132  Cyanobacterial communities from the Joint Geno...   \n",
       "4    112340  Thawing permafrost microbial communities from ...   \n",
       "\n",
       "   its_proposal_id ncbi_project_id                         add_date  \\\n",
       "0              NaN            None  12-MAY-15 12.00.00.000000000 AM   \n",
       "1           1782.0            None  05-FEB-15 12.00.00.000000000 AM   \n",
       "2           1472.0            None  11-NOV-13 12.00.00.000000000 AM   \n",
       "3           1393.0            None  13-DEC-13 12.00.00.000000000 AM   \n",
       "4           1778.0            None  24-OCT-14 12.00.00.000000000 AM   \n",
       "\n",
       "                          mod_date  contact_id  last_mod_by is_public  \\\n",
       "0  08-JAN-16 12.00.00.000000000 AM      113389     102151.0       Yes   \n",
       "1                             None      102898          NaN       Yes   \n",
       "2  29-MAY-14 12.00.00.000000000 AM      108484     103865.0       Yes   \n",
       "3  07-FEB-14 12.00.00.000000000 AM      100043     103865.0       Yes   \n",
       "4  31-OCT-14 12.00.00.000000000 AM      110735     105383.0       Yes   \n",
       "\n",
       "  gpts_proposal_id  ...  public_dap_count admin_dap_count  \\\n",
       "0             None  ...                 0               7   \n",
       "1             None  ...                 0               1   \n",
       "2             None  ...                 0               0   \n",
       "3             None  ...                 0              13   \n",
       "4             None  ...                 0               0   \n",
       "\n",
       "  public_biosample_count admin_biosample_count is_geba is_hmp  \\\n",
       "0                     50                    50      No     No   \n",
       "1                     24                    24      No     No   \n",
       "2                     26                    26      No     No   \n",
       "3                      8                     8      No     No   \n",
       "4                      6                     6      No     No   \n",
       "\n",
       "  ecosystem_path_id is_test  principal_investigator_name               doi  \n",
       "0              4165      No                 James Stegen  10.25585/1487765  \n",
       "1              4234      No                  Byron Crump  10.25585/1487766  \n",
       "2              3948      No               Mary Firestone  10.25585/1487760  \n",
       "3              4318      No                Matthias Hess  10.25585/1487758  \n",
       "4              4206      No                Virginia Rich  10.25585/1487764  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study.head() # peek at data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_dictdf = study.to_dict(orient=\"records\") # transorm dataframe to dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## print out a single record for viewing\n",
    "# for record in study_dictdf:\n",
    "#     print(json.dumps(record, indent=4)); break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## specify characteristics\n",
    "characteristics = \\\n",
    "    ['gold_study_name', 'principal_investigator_name', 'add_date', 'mod_date', 'doi',\n",
    "      'ecosystem', 'ecosystem_category', 'ecosystem_type', 'ecosystem_subtype', 'specific_ecosystem', 'ecosystem_path_id']\n",
    "\n",
    "## create list of json string objects\n",
    "study_json_list = dop.make_json_string_list\\\n",
    "    (study_dictdf, nmdc.Study, id_key='gold_id', name_key='study_name', description_key=\"description\", characteristic_fields=characteristics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"id\": \"Gs0114663\",\n",
      "   \"name\": \"Groundwater microbial communities from the Columbia River, Washington, USA\",\n",
      "   \"description\": \"A metagenomic study to couple microbial communities to carbon and contaminant biogeochemistry in the groundwater-surface water interaction zone\",\n",
      "   \"annotations\": [\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"add_date\"\n",
      "         },\n",
      "         \"has_raw_value\": \"12-MAY-15 12.00.00.000000000 AM\"\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"mod_date\"\n",
      "         },\n",
      "         \"has_raw_value\": \"08-JAN-16 12.00.00.000000000 AM\"\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"gold_study_name\"\n",
      "         },\n",
      "         \"has_raw_value\": \"Microbial communities from the groundwater-surface water interaction zone near Columbia River, Washington, USA and their role in carbon and contaminant biogeochemistry\"\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"ecosystem\"\n",
      "         },\n",
      "         \"has_raw_value\": \"Environmental\"\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"ecosystem_category\"\n",
      "         },\n",
      "         \"has_raw_value\": \"Aquatic\"\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"ecosystem_type\"\n",
      "         },\n",
      "         \"has_raw_value\": \"Freshwater\"\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"ecosystem_subtype\"\n",
      "         },\n",
      "         \"has_raw_value\": \"Groundwater\"\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"specific_ecosystem\"\n",
      "         },\n",
      "         \"has_raw_value\": \"Unclassified\"\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"ecosystem_path_id\"\n",
      "         },\n",
      "         \"has_raw_value\": 4165\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"principal_investigator_name\"\n",
      "         },\n",
      "         \"has_raw_value\": \"James Stegen\"\n",
      "      },\n",
      "      {\n",
      "         \"has_characteristic\": {\n",
      "            \"name\": \"doi\"\n",
      "         },\n",
      "         \"has_raw_value\": \"10.25585/1487765\"\n",
      "      }\n",
      "   ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(study_json_list[0]) ## peek at data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "dop.save_json_string_list(\"output/nmdc-json/study.json\", study_json_list) # save json string list to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Buid project json\n",
    "* Create dataframe for projects that are part of the FICUS studies.\n",
    "* Follow approach for building json data files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    project.gold_id, project.project_name, project.description, project.add_date, \n",
    "    project.mod_date, project.completion_date, project.ncbi_project_name, study.gold_id as study_gold_id\n",
    "from \n",
    "    project\n",
    "inner join study\n",
    "    on study.study_id = project.master_study_id\n",
    "\"\"\"\n",
    "projectdf = sqldf(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dictdf = projectdf.to_dict(orient=\"records\") # transorm dataframe to dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## specify characteristics\n",
    "characteristics = \\\n",
    "    ['add_date', 'mod_date', 'completion_date', 'ncbi_project_name']\n",
    "\n",
    "## create list of json string objects\n",
    "project_json_list = dop.make_json_string_list\\\n",
    "    (project_dictdf, nmdc.SequencingProject, id_key='gold_id', name_key='project_name', \n",
    "     part_of_key=\"study_gold_id\", description_key=\"description\", characteristic_fields=characteristics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(project_json_list[0]) ## peek at data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "dop.save_json_string_list(\"output/nmdc-json/project.json\", project_json_list) # save json string list to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build biosample json\n",
    "* Create dataframe for biosamples that are part of the FICUS studies. Note the use of `group_concat` in the query string. This is needed b/c a biosample may potentially belong to more than one project.\n",
    "* Follow approach for building json data files.\n",
    "\n",
    "**Note:** The list of characteristics is quite long. I might need to rething a more elegant way to do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    biosample.gold_id,\n",
    "    biosample.biosample_name,\n",
    "    biosample.description,\n",
    "    biosample.add_date,\n",
    "    biosample.mod_date,\n",
    "    biosample.ecosystem_path_id,\n",
    "    biosample.ecosystem,\n",
    "    biosample.ecosystem_category,\n",
    "    biosample.ecosystem_type,\n",
    "    biosample.ecosystem_subtype,\n",
    "    biosample.specific_ecosystem,\n",
    "    biosample.habitat,\n",
    "    biosample.location,\n",
    "    biosample.community,\n",
    "    biosample.ncbi_taxonomy_name,\n",
    "    biosample.geographic_location,\n",
    "    biosample.latitude,\n",
    "    biosample.longitude,\n",
    "    biosample.sample_collection_site,\n",
    "    biosample.identifier,\n",
    "    biosample.sample_collection_year,\n",
    "    biosample.sample_collection_month,\n",
    "    biosample.sample_collection_day,\n",
    "    biosample.sample_collection_hour,\n",
    "    biosample.sample_collection_minute,\n",
    "    biosample.host_name,\n",
    "    biosample.depth,\n",
    "    biosample.subsurface_depth,\n",
    "    biosample.altitude,\n",
    "    biosample.temperature_range,\n",
    "    biosample.proport_woa_temperature,\n",
    "    biosample.biogas_temperature,\n",
    "    biosample.growth_temperature,\n",
    "    biosample.soil_annual_season_temp,\n",
    "    biosample.water_samp_store_temp,\n",
    "    biosample.biogas_retention_time,\n",
    "    biosample.salinity,\n",
    "    biosample.pressure,\n",
    "    biosample.ph,\n",
    "    biosample.chlorophyll_concentration,\n",
    "    biosample.nitrate_concentration,\n",
    "    biosample.oxygen_concentration,\n",
    "    biosample.salinity_concentration,\n",
    "    group_concat(project.gold_id) as project_gold_ids\n",
    "from\n",
    "    biosample\n",
    "inner join project_biosample\n",
    "    on biosample.biosample_id = project_biosample.biosample_id\n",
    "inner join project\n",
    "    on project.project_id = project_biosample.project_id\n",
    "group by\n",
    "    biosample.biosample_id,\n",
    "    biosample.biosample_name,\n",
    "    biosample.description,\n",
    "    biosample.add_date,\n",
    "    biosample.mod_date,\n",
    "    biosample.ecosystem_path_id,\n",
    "    biosample.ecosystem,\n",
    "    biosample.ecosystem_category,\n",
    "    biosample.ecosystem_type,\n",
    "    biosample.ecosystem_subtype,\n",
    "    biosample.specific_ecosystem,\n",
    "    biosample.habitat,\n",
    "    biosample.location,\n",
    "    biosample.community,\n",
    "    biosample.ncbi_taxonomy_name,\n",
    "    biosample.geographic_location,\n",
    "    biosample.latitude,\n",
    "    biosample.longitude,\n",
    "    biosample.sample_collection_site,\n",
    "    biosample.identifier,\n",
    "    biosample.sample_collection_year,\n",
    "    biosample.sample_collection_month,\n",
    "    biosample.sample_collection_day,\n",
    "    biosample.sample_collection_hour,\n",
    "    biosample.sample_collection_minute,\n",
    "    biosample.host_name,\n",
    "    biosample.depth,\n",
    "    biosample.subsurface_depth,\n",
    "    biosample.altitude,\n",
    "    biosample.temperature_range,\n",
    "    biosample.proport_woa_temperature,\n",
    "    biosample.biogas_temperature,\n",
    "    biosample.growth_temperature,\n",
    "    biosample.soil_annual_season_temp,\n",
    "    biosample.water_samp_store_temp,\n",
    "    biosample.biogas_retention_time,\n",
    "    biosample.salinity,\n",
    "    biosample.pressure,\n",
    "    biosample.ph,\n",
    "    biosample.chlorophyll_concentration,\n",
    "    biosample.nitrate_concentration,\n",
    "    biosample.oxygen_concentration,\n",
    "    biosample.salinity_concentration\n",
    "\"\"\"\n",
    "biosampledf = sqldf(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_dictdf = biosampledf.to_dict(orient=\"records\") # transorm dataframe to dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "## specify characteristics\n",
    "characteristics = \\\n",
    "    ['add_date',\n",
    "     'mod_date',\n",
    "     'ecosystem_path_id',\n",
    "     'ecosystem',\n",
    "     'ecosystem_category',\n",
    "     'ecosystem_type',\n",
    "     'ecosystem_subtype',\n",
    "     'specific_ecosystem',\n",
    "     'habitat',\n",
    "     'location',\n",
    "     'community',\n",
    "     'ncbi_taxonomy_name',\n",
    "     'geographic_location',\n",
    "     'latitude',\n",
    "     'longitude',\n",
    "     'sample_collection_site',\n",
    "     'identifier',\n",
    "     'sample_collection_year',\n",
    "     'sample_collection_month',\n",
    "     'sample_collection_day',\n",
    "     'sample_collection_hour',\n",
    "     'sample_collection_minute',\n",
    "     'host_name',\n",
    "     'depth',\n",
    "     'subsurface_depth',\n",
    "     'altitude',\n",
    "     'temperature_range',\n",
    "     'proport_woa_temperature',\n",
    "     'biogas_temperature',\n",
    "     'growth_temperature',\n",
    "     'soil_annual_season_temp',\n",
    "     'water_samp_store_temp',\n",
    "     'biogas_retention_time',\n",
    "     'salinity',\n",
    "     'pressure',\n",
    "     'ph',\n",
    "     'chlorophyll_concentration',\n",
    "     'nitrate_concentration',\n",
    "     'oxygen_concentration',\n",
    "     'salinity_concentration'\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "## create list of json string objects\n",
    "biosample_json_list = dop.make_json_string_list\\\n",
    "    (biosample_dictdf, nmdc.Biosample, id_key='gold_id', name_key='biosample_name', \n",
    "     part_of_key=\"project_gold_ids\", description_key=\"description\", characteristic_fields=characteristics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(biosample_json_list[0]) # peek at data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "dop.save_json_string_list(\"output/nmdc-json/biosample.json\", biosample_json_list) # save json string list to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
